modules/apps/segments/segments-web/src/main/java/com/liferay/segments/web/internal/display/context/EditSegmentsEntryDisplayContext.java

```diff
@@ -374,25 +377,40 @@ private JSONObject _getInitialQueryJSONObject(
 		FilterParser filterParser = _filterParserProvider.provide(
 			segmentsCriteriaContributor.getEntityModel());
 
-		Expression expression = filterParser.parse(criterionFilterString);
+		try {
+			Expression expression = filterParser.parse(criterionFilterString);
+
+			JSONObject jsonObject = (JSONObject)expression.accept(
+				new ExpressionVisitorImpl(
+					1, segmentsCriteriaContributor.getEntityModel()));
+
+			if (Validator.isNull(jsonObject.getString("groupId"))) {
+				jsonObject = JSONUtil.put(
+					"conjunctionName",
+					StringUtil.toLowerCase(
+						String.valueOf(BinaryExpression.Operation.AND))
+				).put(
+					"groupId", "group_0"
+				).put(
+					"items", JSONUtil.putAll(jsonObject)
+				);
+			}
 
-		JSONObject jsonObject = (JSONObject)expression.accept(
-			new ExpressionVisitorImpl(
-				1, segmentsCriteriaContributor.getEntityModel()));
+			return jsonObject;
+		}
+		catch (ExpressionVisitException expressionVisitException) {
+			if (_log.isDebugEnabled()) {
+				_log.error(expressionVisitException);
+			}
 
-		if (Validator.isNull(jsonObject.getString("groupId"))) {
-			jsonObject = JSONUtil.put(
-				"conjunctionName",
-				StringUtil.toLowerCase(
-					String.valueOf(BinaryExpression.Operation.AND))
-			).put(
-				"groupId", "group_0"
-			).put(
-				"items", JSONUtil.putAll(jsonObject)
-			);
+			if (!GetterUtil.getBoolean(
+					PropsUtil.get("feature.flag.LPS-166954"))) {
+
+				throw expressionVisitException;
+			}
 		}
 
-		return jsonObject;
+		return JSONUtil.put("invalidSegmentError", StringPool.TRUE);
```

Comment:
But Why this segment is invalid?

What are exactly the errors you are trying to handle?

Reply:
In case the Exception is catch and the FF is true, we're adding that "invalidSegmentError". The error is called "Invalid Segment Error" in the story, but as in the story says, right now we're handling just errors in a segment when we delete a Custom Field which was used by that segment. In case FF if false, we're throwing the Exception and the old Segment Invalid Error is displayed.